File: hydra-python/src/main/python/hydra/lib/sets.py
"""Python implementations of hydra.lib.sets primitives."""

from __future__ import annotations
from collections.abc import Callable, Sequence
from typing import Any, TypeVar

from hydra.dsl.python import frozenlist

A = TypeVar('A')
B = TypeVar('B')


def delete(x: A, s: frozenset[A]) -> frozenset[A]:
    """Delete an element from a set."""
    return s - {x}


def difference(s1: frozenset[A], s2: frozenset[A]) -> frozenset[A]:
    """Compute the difference of two sets."""
    return s1 - s2


def empty() -> frozenset[Any]:
    """Create an empty set."""
    return frozenset()


def from_list(xs: Sequence[A]) -> frozenset[A]:
    """Create a set from a list."""
    return frozenset(xs)


def insert(x: A, s: frozenset[A]) -> frozenset[A]:
    """Insert an element into a set."""
    return s | {x}


def intersection(s1: frozenset[A], s2: frozenset[A]) -> frozenset[A]:
    """Compute the intersection of two sets."""
    return s1 & s2


def map(f: Callable[[A], B], s: frozenset[A]) -> frozenset[B]:
    """Map a function over a set."""
    return frozenset(f(x) for x in s)


def member(x: A, s: frozenset[A]) -> bool:
    """Check if an element is in a set."""
    return x in s


def null(s: frozenset[Any]) -> bool:
    """Check if a set is empty."""
    return len(s) == 0


def singleton(x: A) -> frozenset[A]:
    """Create a singleton set."""
    return frozenset({x})


def size(s: frozenset[Any]) -> int:
    """Get the size of a set."""
    return len(s)


# TODO: ensure that Hydra's native comparison primitives are used for sorting
def to_list(s: frozenset[A]) -> frozenlist[A]:
    """Convert a set to a list.

    Attempts to sort if elements are comparable, otherwise returns unsorted.
    """
    try:
        return tuple(sorted(s))  # type: ignore[type-var]
    except TypeError:
        # Elements don't support comparison (e.g., Term objects)
        return tuple(s)


def union(s1: frozenset[A], s2: frozenset[A]) -> frozenset[A]:
    """Compute the union of two sets."""
    return s1 | s2


def unions(sets: Sequence[frozenset[A]]) -> frozenset[A]:
    """Compute the union of multiple sets."""
    result = frozenset[A]()
    for s in sets:
        result = result | s
    return result

File: overlay/go/hydra-kernel/src/main/go/hydra/core/overlay/go/lib/sets/sets.go
// Package sets implements the hydra.core.lib.sets primitives over the persistent,
// canonical-order util.Set. Sets are immutable; every mutator returns a new set.
package sets

import "hydra.dev/hydra/overlay/go/util"

func asSet(x any) util.Set { return x.(util.Set) }

// Delete : a -> set<a> -> set<a>
func Delete(x any, s any) any { return asSet(s).Delete(x) }

// Difference : set<a> -> set<a> -> set<a> (elements of a not in b)
func Difference(a any, b any) any {
	bs := asSet(b)
	res := util.EmptySet()
	for _, x := range asSet(a).ToList() {
		if !bs.Member(x) {
			res = res.Insert(x)
		}
	}
	return res
}

// Empty : set<a>
func Empty() any { return util.EmptySet() }

// Filter : (a -> boolean) -> set<a> -> set<a>
func Filter(p any, s any) any {
	pred := p.(func(any) bool)
	out := []any{}
	for _, x := range asSet(s).ToList() {
		if pred(x) {
			out = append(out, x)
		}
	}
	return util.NewSet(out...)
}

// FromList : list<a> -> set<a>
func FromList(lst any) any { return util.NewSet(lst.([]any)...) }

// Insert : a -> set<a> -> set<a>
func Insert(x any, s any) any { return asSet(s).Insert(x) }

// Intersection : set<a> -> set<a> -> set<a>
func Intersection(a any, b any) any {
	bs := asSet(b)
	res := util.EmptySet()
	for _, x := range asSet(a).ToList() {
		if bs.Member(x) {
			res = res.Insert(x)
		}
	}
	return res
}

// Map : (a -> b) -> set<a> -> set<b>
func Map(f any, s any) any {
	g := f.(func(any) any)
	out := []any{}
	for _, x := range asSet(s).ToList() {
		out = append(out, g(x))
	}
	return util.NewSet(out...)
}

// Member : a -> set<a> -> boolean
func Member(x any, s any) any { return asSet(s).Member(x) }

// IsEmpty : set<a> -> boolean
func IsEmpty(s any) any { return asSet(s).Size() == 0 }

// Singleton : a -> set<a>
func Singleton(x any) any { return util.NewSet(x) }

// Size : set<a> -> int32
func Size(s any) any { return int32(asSet(s).Size()) }

// ToList : set<a> -> list<a>
func ToList(s any) any { return asSet(s).ToList() }

// Union : set<a> -> set<a> -> set<a>
func Union(a any, b any) any {
	res := asSet(b)
	for _, x := range asSet(a).ToList() {
		res = res.Insert(x)
	}
	return res
}

// Unions : list<set<a>> -> set<a>
func Unions(lst any) any {
	res := util.EmptySet()
	for _, s := range lst.([]any) {
		for _, x := range asSet(s).ToList() {
			res = res.Insert(x)
		}
	}
	return res
}
